### ISSUE TITLE
Proposal: make eject non-destructive

### ISSUE THREAD

**Issue by algal** [mxcl/swift-sh#52]
H! It is not correct to call this an "issue." This is really an enhancement to the way that the `eject` command works, which I wanted to propose for discussion. I'm guessing this is the right place to raise the issue?  So here goes.

## introduction

Right now, `eject` does two things: first, it erases your existing swift-sh script, and second, it creates a new Swift Package Manager (SPM) project based on the script.

I think one usually wants the second thing, creating a new project.

So I'd propose that `eject` should do this by default. Or else, it should take an optional flag `--nondestructive` which makes it work this way, creating the SPM project in its own directory but not erasing the original script.

## motivation

Why? A few reasons.

First, `eject` feels like a compilation tool. It takes an input (a single file source script) and generates an output (a folder with a complete SPM project). It would be more intuitive and familiar if it behaved like other compilation tools, which produce an output but do not destroy their input along the way. This is how `swiftc`, `clang`, `gcc`, `markdown.pl` -- heck, even plain old `zip`-- all behave, and I think it is the right default. This is not only because of convention but also  because it gives the user more options regarding what to do next.

Second, I have found I _usually_ want to hold onto the original script because of the workflow I find I am using.

This is what you might call the "pop into Xcode" workflow. Working on a single file script, I want to jump into Xcode for a few hours, basically just to have the benefit of autocomplete and documentation quick lookup, over the standard libraries, my dependencies, and my own code. So I use `eject` to create an SPM project, and I run `swift package generate-xcodeproj` to generate the Xcode project, and then I work in Xcode. But then once I'm done coding, I don't want or need the complexity of an Xcode project or of an SPM project. I want to go back to the simplicity of a single file.

I also find `swift-sh` is easier than `Package.swift` for specifying dependencies. This is hard in `Package.swift` is hard, because of the subtle way `Package.swift` mixes nested enums, static functions, and argument overloads in order to define the specification, and because of the need to specify target relationships correctly. The `swift-sh` syntax for specifying dependency as an in-line comments is much easier, so in this area switching to SPM and to Xcode is a step backward. I'd rather iterate on my dependencies in `swift-sh`.

When I'm done in Xcode, I basically manually "reverse-eject" back into a swift-sh script. This is easier to do if I am just copying the new code into the original script file, rather than recreating a file that has been deleted.

Essentially, I think it would make sense for `eject` to be destructive _only if_ the expected workflow were as follows: You start by hacking on a swift-sh script for a while, as a single file, using a plain old text editor. Then, eventually, when it gets complicated enough, you finally decide to convert it to a SPM project and maybe to an Xcode project. This is a one way transition and you will probably never want to go back to a single file again, because SPM or Xcode provides features you absolutely need for the project's level of complexity. What are these features, potentially? For instance: organizing code into separate files, setting up separate unit tests, managing associated files like required assets, multiple build targets, IDE autocomplete, multiple build schemes, more fine-grained control and UI for build settings, etc..

But what I find is that most of the time I want to "eject" to get to Xcode just for a couple of conveniences, temporarily for workflow benefits, and not as a sign of permanent transition in the project's complexity. This kind of "temporary eject" is likely to happen lots of times, but the "permanent eject" happens only once. So `swift-sh` should default to facilitating the more common case.

## proposal

I see two options: 

1. Modify `eject` to be non-destructive by default

So if you had an Swift-sh file `hello.swift`, and you did `swift sh eject hello.swift`, it would still create a directory `Hello` as it does now. It just would not delete `hello.swift`

What if you had an executable Swift file named `hello` (which used the `#!/usr/bin/swift sh` shebang to be an executable)? Then you couldn't create a directory named `Hello` because of name conflict. In that case, I'd suggest that `swift sh eject` should just exit with an error complaining about the name conflict. Alternatively, it could take a `-o` or `--output` command line argument and let the user explicitly specify the name of the directory to be created.

2. Modify `eject` to be a non-destructive when invoked with an optional flag `--nondesctructive` to make this behavior optional.

## compatibility

Option 1 would be a breaking change. If anyone out there has already started building scripts that rely on the current destructive `eject` behavior, then this tools would need to be updated to be aware of the new behavior.

Option 2 would be a non-breaking change, since it introduces an optional flag.

Option 1 is better, in my opinion. Best to make the change now before too many folks get used to the wrong thing. 😉

## last words

I realize this is a lot of words for a simple idea but I wanted to be thoughtful about how I suggested this. I expect the work to implement this change take less time than just articulating this proposal, but that's probably how it should be. I think `swift-sh` has an excellent power/weight ratio right now, and provides the same kind of benefit to its users. I'm trying to stay in that design territory.

**Comment by alexito4** [comment]
Hi, to be honest the necessity of generating an Xcode project to edit the script is the main reason I use marathon and its `marathon edit ...`. Maybe swift-sh could do something like that? Hopefully eventually with the work going on the LSP we can have the niceties of Xcode in other editors without the need of creating a project file.

**Comment by mxcl** [comment]
Seems like really you just want to be able to open swift-sh scripts in Xcode…?

**Comment by algal** [comment]
Not exactly. I want to be able to edit and run swift-sh script in Xcode, because I want to get the benefit of the following things: Xcode’s syntax highlighting, live type checking, fix it suggestions, quick lookup of documentation, breakpoints and debugger integration, and jump-to-definition functionality.

**Comment by mxcl** [comment]
I don’t see how this is different to what I said.

**Comment by algal** [comment]
The difference I had in mind was just that to get those Xcode benefits you need to edit the script in the context of an entire project (rather than doing something like `open -a "Xcode.app" /path/to/file`, which merely uses Xcode as text editor). But, sorry, I was replying off the cuff, and hadn't defined `edit` so it is editing the script in the cached project context.

This is exactly what I was getting at. This `edit` approach definitely seems like a more elegant approach than modifying `eject`.

However, I'm noticing that when I actually try to `swift sh edit` a file and save in Xcode, I get this error:

<img width="1155" alt="screen shot 2019-02-17 at 17 12 11" src="https://user-images.githubusercontent.com/264412/52922734-47c61780-32d8-11e9-9673-8cf02d4fe97c.png">

Just modifying the file and trying to quit Xcode produces this similar error from failed autosave:

<img width="1155" alt="screen shot 2019-02-17 at 17 12 03" src="https://user-images.githubusercontent.com/264412/52922753-60cec880-32d8-11e9-9480-926ddaed9c8b.png">

I believe this is happening because in the cached Xcode project there's a symlink to the original script, and Xcode doesn't seem to support chasing the symlink and saving to the original file. 

Here are some possible fixes. None is clearly the way to go.

*1. hard links*. I tried replacing the symlink with a hard link. This also seems not to work, because Xcode creates a fresh inode when it saves the file, rather than saving to the same inode entry being shared by with the original script. This is frustrating. Perhaps there is a way to force Xcode to be more tolerant of symlinks or hard links to files? I don't know it.

*2. symlinks to folders*. I've done a little experimenting and it seems Xcode _will_ tolerate symlinks to entire directories, just not files. So you could define an Xcode project that contained a symlink to the directory containing the original script, but then you'd be stuck with the original script's file name, and I believe Xcode allows only `main.swift` to have a shebang line. I am not 100% sure of this. Another major drawback of this approach is that, I presume, it requires something more complicated than running `swift package generate-xcodeproj`. I am not sure if that command supports doing anything clever with symlinks in the SPM directory structure.

*3. sync the edited file*. Yet another workaround could be for `swift sh edit` to wait until Xcode quits, and then copy the changed file from the cached project back over the original script. But this is hardly elegant since it blocks the terminal during the editing session, and incorrectly associates the editing session for that single file with the lifetime of the Xcode process. Also, I'm not even sure how to block the script until Xcode quits.

**Comment by mxcl** [comment]
Ugh. I can’t believe I didn’t try saving the file when making this feature. Will figure out a fix.

**Comment by mxcl** [comment]
Anyway closing this, IMO eject *should* be destructive, it's in the name. You wouldn't expect your ejector seat to autopilot the plane back to the ground after you use it.

Will open a new ticket to fix the symlink issue.

Reopen if you feel strongly still.

**Comment by algal** [comment]
I agree an `edit` command is a better way to meet the requirements I was trying to meet with a non-destructive eject.

**Comment by mxcl** [comment]
Fixed #58
